File: kb_spark/Scala/chapter2/ch2_authors_ages_dataframe.scala
/**
 * Topic: Structuring Spark: Key merits and benefits
    ~ Expressivity
    ~ Simplicity
    ~ Composiability
    ~ Uniformity

   >> Using high-level DataFrame API
 */

 package main.scala.chapter2

 import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.functions.avg

 object AuthorsAges {
    def main(args: Array[String]) {
        
        val spark = SparkSession
            .builder
            .appName("AuthorsAges")
            .getOrCreate()

        val dataDF = spark.createDataFrame(Seq(("Brooke", 20), ("Denny", 31), 
            ("Jules", 30), ("TD", 35), ("Brooke", 25))).toDF("name", "age")
        
        val avgDF = dataDF.groupBy("name").agg(avg("age"))
        avgDF.show()
    }
 }

File: kb_spark/Python/chapter2/ch2_authors_ages_dataframe.py
''' Topic: Structuring Spark: Key merits and benefits
        ~ Expressivity
        ~ Simplicity
        ~ Composiability
        ~ Uniformity
    
    >> Using high-level DataFrame API
'''

# Demonstrating expressivity and composability

from pyspark.sql import SparkSession
from pyspark.sql.functions import avg


# Create a dataframe using SparkSession
spark = (SparkSession
         .builder
         .appName("AuthorsAges")
         .getOrCreate())

data_df = spark.createDataFrame([("Brooke", 20), ("Denny", 31), ("Jules", 30),
                                 ("TD", 35), ("Brooke", 25)], ["name", "age"])

avg_df = data_df.groupBy("name").agg(avg("age"))
avg_df.show()
